File: 2019 PAC II - AED - 0800 - Semana de Ejercicios - Ejercicio 1/Python/Lista.py
# -*-coding: "utf-8" -*-
#/***********************Nodo********************************/

class Nodo:
    def __init__(self, valor):
        self.valor = valor
        self.siguiente = None


#/*********************Lista Enlazada**************************/

class ListaEnlazada:
    def __init__(self):
        #atributo
        self.primero = None
             
    def agregar(self,valor, posicion = None):

        if(posicion is None):
            return self.agregarNormal(valor, self.primero)
                
        else:
            return self.agregarEnPosicion(valor,posicion)

    
    def agregarNormal(self,valor, actual = None):
        
        if(not self.primero):
            self.primero = Nodo(valor)
            return True
        
        elif(actual.siguiente):
            return self.agregarNormal(valor,actual.siguiente)
    
        else:
            actual.siguiente = Nodo(valor)
            return True
    
        return False

    
    def agregarEnPosicion(self,valor, posicion):
       
        if(posicion == 0):
            cola = self.primero
            self.primero = Nodo(valor)
            self.primero.siguiente = cola
            return True
        
        else:
            if(posicion > self.obtenerLongitud()):
                return self.agregar(valor)
            
            else:
                actual = self.primero
                anterior = None
                posicionActual = 0

                while(actual):
                    if(posicion == posicionActual):
                        anterior.siguiente = Nodo(valor)
                        anterior.siguiente.siguiente = actual
                        return True
                    
                    else:
                        anterior = actual
                        actual = actual.siguiente
                        posicionActual += 1
            return False

    def obtenerLongitud(self):
        contador = 1
        actual = self.primero

        while(actual.siguiente):
            actual = actual.siguiente
            contador += 1
        
        return contador

    def imprimirLista(self):
        camino = ""
        nodoActual = self.primero
    
        while(nodoActual.siguiente):
            camino = ("%s%s -> " %(camino, nodoActual.valor))
            nodoActual = nodoActual.siguiente
        
        camino = ("%s%s -> None" %(camino, nodoActual.valor))
    
        print(camino)
    
    
    

File: 2019 PAC II - AED - 0800 - Semana de Ejercicios - Ejercicio 1/JavaScript/Lista.js
/***********************Nodo********************************/

function Nodo(valor){
    this.valor = valor;
    this.siguiente = null;
}

/*********************Lista Enlazada**************************/

function ListaEnlazada(){
    //atributo
    this.primero = null;

    //metodos
    this.agregar = ListaEnlazadaAgregar;
    this.agregarNormal = ListaEnlazadaAgregarNormal;
    this.agregarEnPosicion = ListaEnlazadaAgregarEnPosicion;
    this.obtenerTamaño = ListaEnlazadaObtenerTamaño;
    this.imprimirLista = ListaEnlazadaImprimir;
}

    function ListaEnlazadaAgregar(valor, posicion = null){
        if(posicion == null){
            return this.agregarNormal(valor);
        }
        else{
            return this.agregarEnPosicion(valor, posicion);
        }
    }

    function ListaEnlazadaAgregarNormal(valor, actual = this.primero){
        if(!this.primero){
            this.primero = new Nodo(valor);
            return true;
        }

        else{
            if(actual.siguiente){
                return this.agregarNormal(valor, actual.siguiente);
            }

            else{
                actual.siguiente = new Nodo(valor);
                return true;
            }
        }
        return false;
    }

    function ListaEnlazadaAgregarEnPosicion(valor, posicion){
        if(posicion == 0){
            cola = this.primero;
            this.primero = new Nodo(valor);
            this.primero.siguiente = cola;
            return true;
        }
        else{
            if(posicion > this.obtenerTamaño()){
                return this.agregar(valor);
            }
            else{
                actual = this.primero;
                anterior = null;
                posicionActual = 0;

                while(actual){
                    if(posicion == posicionActual){
                        anterior.siguiente = new Nodo(valor);
                        anterior.siguiente.siguiente = actual;
                        return true;
                    }
                    else{
                        anterior = actual;
                        actual = actual.siguiente;
                        posicionActual++;
                    }
                }
            }
            return false;
        }
    }

    function ListaEnlazadaObtenerTamaño(){
        contador = 1;
        actual = this.primero;

        while(actual.siguiente){
            actual = actual.siguiente;
            contador++;
        }

        return contador;
    }

   function ListaEnlazadaImprimir(){
        camino = " ";
    
        nodoActual = this.primero;
    
        while(nodoActual.siguiente){
            camino = camino + nodoActual.valor + " -> ";
    
            nodoActual = nodoActual.siguiente;
        }
    
        camino = camino + nodoActual.valor + " -> NULL";
    
        console.log(camino);
    }